clang/lib/CIR/Dialect/Transforms/CIRSimplify.cpp

```diff
@@ -107,6 +111,110 @@ struct RemoveTrivialTry : public OpRewritePattern<TryOp> {
   }
 };
 
+/// Simplify suitable ternary operations into select operations.
+///
+/// Only those ternary operations that meet the following criteria can be
+/// simplified:
+/// - The true branch and the false branch cannot have any side effects;
+/// - The true branch and the false branch cannot be "too costly" since both of
+///   them will be executed after the folding happens.
+///
+/// For now we only simplify those ternary operations whose true and false
+/// branches either directly yield a value, or directly yield a constant, or
+/// directly yield a value loaded from an alloca. That is, both of the two
+/// branches of these ternary operation must either:
+///
+/// - Only contain a single cir.yield operation, or
+/// - Contain a cir.const operation followed by a cir.yield operation that
+///   yields the constant value produced by the cir.const operation, or
+/// - Contain a cir.load operation followed by a cir.yield operation that yields
+///   the value produced by the cir.const operation. The cir.load operation must
+///   load a non-volatile object from an alloca.
+///
+/// For example, we will simplify the following ternary operation:
+///
+///   %0 = cir.ternary (%condition, true {
+///     %1 = cir.const ...
+///     cir.yield %1
+///   } false {
+///     %2 = cir.load ...
+///     cir.yield %2
+///   })
+///
+/// into the following sequence of operations:
+///
+///   %1 = cir.const ...
+///   %2 = cir.load ...
+///   %0 = cir.select if %condition then %1 else %2
+struct SimplifyTernary final : public OpRewritePattern<TernaryOp> {
+  using OpRewritePattern<TernaryOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(TernaryOp op,
+                                PatternRewriter &rewriter) const override {
+    llvm::SmallVector<mlir::Operation *> opsToHoist;
+
+    mlir::Value trueValue =
+        simplifyTernaryBranch(op.getTrueRegion(), opsToHoist);
+    if (!trueValue)
+      return mlir::failure();
+
+    mlir::Value falseValue =
+        simplifyTernaryBranch(op.getFalseRegion(), opsToHoist);
+    if (!falseValue)
+      return mlir::failure();
+
+    for (auto *hoistOp : opsToHoist)
+      rewriter.moveOpBefore(hoistOp, op);
+    rewriter.replaceOpWithNewOp<mlir::cir::SelectOp>(op, op.getCond(),
+                                                     trueValue, falseValue);
+
+    return mlir::success();
+  }
+
+private:
+  mlir::Value simplifyTernaryBranch(
+      mlir::Region &region,
+      llvm::SmallVector<mlir::Operation *> &opsToHoist) const {
+    if (!region.hasOneBlock())
+      return nullptr;
+
+    mlir::Block &block = region.front();
+
+    // The block can contain at most 2 operations
+    if (block.getOperations().size() > 2)
+      return nullptr;
```

Comment:
Checking the number of operations is not ideal, we could be a bit more generic: if cir.yield result comes out of a constant, inline the current region in the previous block.

Reply:
This might also not work perfectly, I'm actually fine with checking for just a yield + constant and making sure there's nothing else in the region/block.

Reply:
Updated.